In the migration, install and activate Genesis Custom Blocks

kienstra opened the pull request:
#### Changes
* As John suggested, changes the order of the GCB installation and block migration
* Installs and activates GCB via a REST API endpoint
* The endpoint is mainly taken from the Gutenberg [plugin installation route](https://github.com/WordPress/gutenberg/blob/8d64aa3092d5d9e841895bf2d495565c9a770238/lib/class-wp-rest-plugins-controller.php#L417)
* Ideally the JS file would just use that endpoint, but it's only in Gutenberg, not in Core

#### Testing instructions (updated 23 July)

1. Checkout this PR's branch, and also checkout https://github.com/studiopress/genesis-custom-blocks/pull/13, but don't activate GCB it yet
2. Go to `/wp-admin` > Block Lab > Migrate, and go through the process
3. Expected: 
![expected-diff](https://user-images.githubusercontent.com/4063887/88240490-e765a300-cc4c-11ea-83ff-bb40ffa6c0f4.gif)

4. Change the slug to `amp`, as that's available on wp.org to download:

```diff
diff --git a/php/admin/migration/class-api.php b/php/admin/migration/class-api.php
index 9d4b3ae..7e36d98 100644
--- a/php/admin/migration/class-api.php
+++ b/php/admin/migration/class-api.php
@@ -45,7 +45,7 @@ class Api extends Component_Abstract {
         * Api constructor.
         */
        public function __construct() {
-               $this->new_plugin_slug = 'genesis-custom-blocks';
+               $this->new_plugin_slug = 'amp';
                $this->new_plugin_file = "{$this->new_plugin_slug}/{$this->new_plugin_slug}.php";
        }
```
GCB is not on wp.org of course, so we can't really test downloading and installing it.

5. Go through the migration flow just like before
6. Expected: The plugin should have been installed and activated:

<img width="726" alt="plugin-activated" src="https://user-images.githubusercontent.com/4063887/88241075-ab334200-cc4e-11ea-9b33-df3b6c1ff08b.png">

7. To reverse the migration, you might need to do this:
```sh
wp shell
( new Block_Lab\Admin\Migration\Post_Content( 'amp', 'block-lab' ) )->migrate_all();
```

8. Here are more steps to [reverse the migration](https://github.com/getblocklab/block-lab/pull/629#issuecomment-655846402)

9. If you have time, testing [these steps](https://github.com/getblocklab/block-lab/pull/629#issue-439492116) again would help, verifying the 'Expected' steps 8 and 9

Thanks for your patience 😄

kienstra committed:
Mainly use the plugin installation logic from Gutenberg

It has a route to install plugins:
https://github.com/WordPress/gutenberg/blob/8d64aa3092d5d9e841895bf2d495565c9a770238/lib/class-wp-rest-plugins-controller.php#L417

mindctrl commented:
One thing I just thought about with this migration flow. The activate plugin step is last. That means if that step fails, the customer's site is left in a broken state. Is there a reason for not installing the new plugin first? I saw mention of bootstrapping conflicts. Is that difficult to solve here? We had a similar issue in AB with AB Pro/GPB, but solved it in the main plugin file by moving most of the logic to a secondary loader file.

Even with the conflicts resolved, this still assumes the plugin will install and activate successfully _after_ the content is migrated. It would be nice to verify that you can complete the required steps before updating all the content.

kienstra commented:
Hi @mindctrl,
> Is there a reason for not installing the new plugin first? I saw mention of bootstrapping conflicts. Is that difficult to solve here? 

Great idea. That should be possible with https://github.com/studiopress/genesis-custom-blocks/pull/11. I'll look at it.

kienstra commented:
Hi @mindctrl 
What do you think about this?

1. Exchange steps 4 and 5
2. The step 'Install And Activate Genesis Custom Blocks' would install and activate the plugin with the REST API. So there'd be no reload. If either failed, we could prevent migrating in the next step:
<img width="651" alt="exchanging-steps-here" src="https://user-images.githubusercontent.com/4063887/88091699-0b45bd80-cb55-11ea-92b5-9ffad54480f8.png">

3. The 'Migrate your Blocks' step would still have the CPT and `post_content` migration. But it wouldn't install GCB anymore. On success, it would have a button 'Go to Genesis Custom Blocks' that would deactivate BL and go to `/wp-admin` > Custom Blocks:

<img width="821" alt="migration-here" src="https://user-images.githubusercontent.com/4063887/88091612-ed785880-cb54-11ea-8121-1298fb24cdbf.png">

mindctrl commented:
Hi @kienstra 

> If either failed, we could prevent migrating in the next step:

I like the idea of changing the order and confirming GCB can be and is successfully installed before migrating. That's probably the safest thing to do I think.

kienstra commented:
Thanks, John! Good idea to check the plugin can be activated first.

kienstra commented:
Hi @lukecarbis and @RobStino,
What do you think about https://github.com/getblocklab/block-lab/pull/653#issuecomment-662029626? 

As John brought up, it could prevent users from having errors in blocks if the plugin installation and activation don't work.

lukecarbis commented:
Great idea!

kienstra commented:
Nice, thanks! I'll apply that tomorrow.

kienstra committed:
Display the plugin conflict notice in Block Lab, not GCB

As Luke sugggested, use the 'plugins_loaded' hook
So GCB will have loaded the helper functions,
and this look for them with function_exists(),
then prevent loading them and display a notice.
The notice will only be on plugin and Block Lab pages,
not everywhere in /wp-admin.

kienstra committed:
Add a 'Deactivate' link to deactivate Block Lab

As part of the plugin conflict notice,
this will only appear if GCB is also active.
Also, move require statements for helper
functions to 'plugins_loaded' hook,
as this will run 1 priority after GCB,
and won't cause a PHP error.

kienstra committed:
Add a test for require_helpers()

If the functions already exist,

It should instead display an admin notice.

kienstra commented:
Here's the notice that Block Lab displays if GCB is also active:

<img width="941" alt="block-lab-displays-notice" src="https://user-images.githubusercontent.com/4063887/88225848-cee78f80-cc30-11ea-8022-047d65f15a5d.png">

>It looks like Block Lab is active. Please deactivate it or migrate, as it will not work while Genesis Custom Blocks is active.

Block Lab won't load its helper functions, like `block_value()` and `block_field()`. But it'll still load all other classes, and all of the menus will work.

kienstra committed:
Implement much of the UI changes John suggested

Move the plugin installation and activation earlier,
and migrate the blocks last.

kienstra committed:
Simplify the const names

They don't all need to have migration,
the could simply have the state.

kienstra committed:
Activate the plugin via the REST API endpoint

Before, the endpoint simply installed the plugin.
Now, it also activates it.
Most of the logic is taken from
Gutenberg.

kienstra committed:
Remove extra file that's not used

kienstra committed:
Speak the error message if there is one

Also, some other fixes

kienstra committed:
Check that the response has error messages before accessing them

This caused a full-page error
with an unexpected response.

kienstra committed:
Rename the route to install-activate-gcb

This is more accurate,
as it's not only installing.

kienstra committed:
After clicking to begin migration, hide button

There's no need for the button
while it's in progress.

kienstra committed:
Fix failed tests in multisite

is_network_only_plugin()
was eventually causing an error,
which also caused other failed tests.

kienstra committed:
Handle a possible error in saving a key

If the REST API is disabled,
handle this with the error message:
'There was an error saving the key.'

kienstra committed:
Change the sub key endpoint to return WP_Error on failure

To follow the convention in the REST API.
Also, add testing.

kienstra committed:
In CPT and post content migration, return WP_Error on failure

Uses standard REST API return on failure.
Also, refactor the migration in the
JS file.

kienstra committed:
Fix return tags in PHP DocBlocs

Also, make the <input> for the
sub key longer.

kienstra committed:
Chain the migration functions

Only run the post content migration
if the first succeeded.

kienstra commented on php/admin/migration/class-post-type.php:
```diff
@@ -84,20 +85,23 @@ public function __construct( $previous_post_type_slug, $previous_block_namespace
+		$errors             = new WP_Error();
 
-		while ( ! empty( $posts ) ) {
+		while ( ! empty( $posts ) && $error_count < $max_allowed_errors ) {
```
Like with the post content, this now exits if there are 20 or more errors.

kienstra replied:
Like you brought up before, if there are several errors, this can be an infinite loop.

kienstra committed:
Remove the escaping of the redirect url

It's not being output,
so it's not needed.

kienstra commented on php/admin/migration/class-submenu.php:
```diff
@@ -36,23 +36,14 @@ class Submenu extends Component_Abstract {
-	 *
-	 * @var string
-	 */
-	const QUERY_VAR_DISABLE_ONBOARDING = 'disable_onboarding';
```
This shouldn't be needed, as this activates GCB 'silently' via `activate_plugin()`, so it won't run the activation hook.

kienstra committed:
Change the format of add_query_arg()

This seems to work also,
but move admin_url() inside of it.

kienstra commented:
Hi @lukecarbis,
Could you please review this when you have a chance? 

Thanks for your patience with all of this 😄

lukecarbis reviewed (approved):
Amazing, as always. Wish I could find something to fault you for.

kienstra commented:
Thanks, Luke! Thanks a lot for your patience with all of the changes here 😄
